### DISCUSSION TITLE
Verovio as a MEI normalizer

### DISCUSSION THREAD

**Discussion by fosfrancesco** [rism-digital/verovio#2953] [Ideas]
Hello community,
I'm working with the problem of parsing MEI files (only common music notation) in Python to run statistics and machine learning on the data that are encoded.
For this goal, the great flexibility of MEI during the encoding phase, becomes a problem to overcome, since the parser needs to check for musical elements in many different formats. 

To give an example (from three files I found on the web), the key signature can be encoded as:
- an element keySig inside staffDef
![image](https://user-images.githubusercontent.com/22941644/177332046-0b9d6edf-852c-491f-aa29-796ed2af7511.png)
- an attribute key.sig inside staffDef
![image](https://user-images.githubusercontent.com/22941644/177332072-d6b813aa-5883-439b-bde7-7cb62fe9a711.png)
- an attribute key.sig inside scoreDef
![image](https://user-images.githubusercontent.com/22941644/177332117-81ff063a-926d-407c-b975-307d9ae925e2.png)

Similar things happen for many other score elements, for example, beamings, accidentals, clefs, etc...
I know some "file normalization" is going on in Verovio under the hood since all these different formats still need to produce the same graphical output.

Would it be possible to make this "normalization" accessible to the user? In this way, I could start my parsing process from a normalized version of the MEI. Also, since many MEI users that write common music notation test their scores using verovio, if I'm able to parse verovio output, the chance that I encounter a score that I can't parse, will be reduced significantly.

In python this could look something like:
1. load the mei file in verovio python toolkit
2. get the "normalized" MEI with .getMEI(normalized=True) function
3. parse that with my library

Thank you for your work and feedback!

**Comment by lpugin** [comment]
A general idea would be to use MEI basic for this, which has, to some extend, this normalization purpose. 

There is [here](https://github.com/rism-digital/verovio/pull/2950) a pull-request that implements MEI basic output. You would do (pseudo-syntax)
```
verovio.getMEI({ 'basic': true })
```
That would remove editorial markup and move attribute to element (or the opposite) as expected for MEI basic.

Regarding the placement of score definitions (i.e., key signatures) in scoreDef vs staffDef, this is not constrained in MEI basic. However, the PR includes an option to move all of them to staffDef when loading the file, so that can be used in conjunction with MEI basic output if you want always the same encoding pattern.

**Comment by fosfrancesco** [comment]
Thank you very much! I will test it out. 

While this is definitely not the main purpose of Verovio, I think this is super useful utility for all people that need to extract information from a large number of MEI files, often written by different people with different styles.
